README.md
# Inject

Inject is a form of pre-processor that replaces/'injects' data from a file
in accordance to data defined in a data map. Inject separates logic 
from the file being read, and because of this, Inject has a variable 
syntax that can adapt to different syntax conventions depending on 
the information it's being used on.

Inject reads the contents of files and maps/replaces data matching the 
specified key/value data specified in the data map. In a nutshell, 
Inject takes any file and turns it into an 'inject' module, where 
the data, the syntax scanned for and the dependencies themselves 
are given externally via dependency injection to Inject so it is 
not specified within the file. The data can then be 'interpreted' 
via the syntax that Inject has been told to use.

## Features

In most programming languages, symbols are used to prefix data (such 
as the '@mixin' syntax) or circumfix data (using a prefix and suffix, 
such as the double curly brace '{{mustache}}' syntax) to 'flag' a 
particular need for processing to be applied to information. However, 
not all do, and the available types of syntax obviously change 
depending on language. With Inject, a prefix or a prefix and suffix 
can be set, but if non are set, it will also work without additional 
flagging on the data, and find the raw triggers in the file (currently 
only a prefix or circumfix can be added to Inject's evaluation; the 
suffix alone is not due to its lack of usage).

The next important feature of Inject is that it's infrastructure has 
been built purely on the use of data streams. By using data streams, 
content can be received before the data from the source had even 
completed processing the entire content. As such, Inject returns a 
stream rather than 'complete' data, but thanks to Dart, the complete 
data can be returned with the use of casting a future from a stream 
if so desired.

Finally, Inject modules have been built to stack on-top of one 
another. For example, if an Inject object is passed as another Inject 
object's data map value, once this value is triggered, the processed 
output from it is injected into the original Inject object's data 
stream. The primary focus for Inject is that it is as modular as 
possible throughout, as such, think of Inject objects as building
blocks that can be stacked on-top of one another and reused due to 
a dependency injected approach. Inject does not require any of the 
files it is used with to have a particular extension, which means a 
HTML file can stay as a .html file, so it can be used as an Inject 
module, or on its own
 
## Usage

Inject can be used to further separate data from logic/processing. 
In an MVC framework for example, Inject can be used as a templating 
engine to keep Views more modular and separate from files that it 
'could' depend on or be used with by specifying this in the injection 
(via a Controller for example), rather than the file itself. This 
makes it less restrictive as to what it can be used with, or what 
additional files it requires, rather than those that 'hard-code' 
dependencies in the file itself.

This first example is the more long and drawn out illustration of 
usage, but this should hopefully articulate it's working better. 
It's taking a class as the data map, thus the fromObject() constructor
is needed.

```dart

import 'package:inject_module/inject_module.dart';
import 'dart:convert';

class User {
  // Example class. Passing this class as an inject data map
  // will mean 'name' and 'age' are the triggers, and their
  // values are what will be injected upon triggering.
  String name = 'Will Squire';
  int age = 25;
}

main() {
  // Specify the location of the file Inject will analyse
  // and process.
  File file = new File('../user_view.html');

  // Define either an Object or a Map to use as Inject's data
  // map. Inject uses the data maps' keys/variable names as
  // the triggers to replace/inject data. Once a trigger is
  // found in the source file's data, its replaced with the
  // data maps' corresponding value for the trigger.
  User user = new User();

  // Inject requires at least two arguments, one is the source
  // file and the other is the data map. Inject can also take
  // another two optional parameters, a prefix and then a suffix.
  // Using a prefix or both a prefix and suffix (circumfix)
  // means inject triggers need these additional significations
  // before triggers can be detected in the data. If using an
  // object as the data map, call fromObject constructor.
  InjectModule injectModule = new InjectModule.fromObject(file, user);

  // Inject can now analyse and transform the given file
  // data, returning the output as a stream.
  injectModule.process()
  .transform(UTF8.decoder)
  .listen((String char) {
    // Do something with the compiled result.
    print(char);
  });
}

```

This next example takes a Map as the data map and injects a 
Stream as the replacement data.

```dart

import 'package:inject_module/inject_module.dart';
import 'dart:convert';
import 'dart:io';

main() {
  // Defining the dataMap as a map type. In this data map the
  // trigger 'body' in will inject the content of a file as a
  // stream. Thus, injecting one stream into another.
  Map<String,Object> dataMap = {
    'body' : new File(new Uri.file('../body.html').toFilePath()).openRead()
  };

  // A third parameters has been passed here, which means
  // triggers will need the prefix of '$' to be recognised, and
  // have the corresponding value 'injected' in its place.
  // (Note, because Dart uses '$' for string interpolation,
  // it has been escaped with a '\').
  InjectModule injectModule = new InjectModule(new File('../base.html'), dataMap, '\$');

  // Process
  injectModule.process()
  .transform(UTF8.decoder)
  .listen((String char) {
    // Result.
    print(char);
  });
}

```

This third and final example of usage presents the stacking 
ability that Inject modules can have on one another. Inject
streams can also inject another Inject stream.

```dart

import 'package:inject_module/inject_module.dart';
import 'dart:convert';

main() {
  // The third and forth parameters have been passed for each
  // Inject module here, which means triggers will need the
  // prefix of '<!--' and suffix of '-->' to be recognised, to
  // have the corresponding value 'injected' in its place.
  // Notice that HTML comment syntax was used for this. This
  // is so the files can be used even without Inject processing
  // the trigger, as it will not show up on the client's
  // machine.
  String prefix = '<!--';
  String suffix = '-->';

  Map<String,Object> headerDataMap = {
    'title' : "Website title"
  };

  Map<String,Object> footerDataMap = {
    'contactNumber' : "0800-Etc"
  };

  // In this data map, the values given are other Inject modules.
  // This takes the output from those streams and injects them
  // at the given location when this stream is processed.
  Map<String,Object> dataMap = {
    'header' : new InjectModule(new File('../header.html'), headerDataMap, prefix, suffix),
    'footer' : new InjectModule(new File('../footer.html'), footerDataMap, prefix, suffix)
  };

  InjectModule injectModule = new InjectModule(new File('../base.html'), dataMap, prefix, suffix);

  // Process
  injectModule.process()
  .transform(UTF8.decoder)
  .listen((String char) {
    // Result.
    print(char);
  });
}

```

## Features and bugs

Please email myself feature requests and bugs, or find me on 
[twitter][Twitter]. It's lovely to hear any kind of feedback, so
feel free to contact me whenever.

[Twitter]: https://twitter.com/WillSquire

CHANGELOG.md
Changelog
---------

1.0.2
=====
### Breaking changes
- Changed Injects' constructor argument to take
a File type instead of URI.

1.0.1
=====
### Bug fixes
- Corrected README.md to correct package import.

1.0.0
=====

- Initial version, created by Will Squire.

test/all_test.dart
// Copyright (c) 2015, Will Squire. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

library inject.test;
/*
import 'package:unittest/unittest.dart';
import 'package:inject/inject_module.dart';

main() {
  group('A group of tests', () {
    Awesome awesome;

    setUp(() {
      awesome = new Awesome();
    });

    test('First Test', () {
      expect(awesome.isAwesome, isTrue);
    });
  });
}
*/

lib/src/inject_base.dart
// Copyright (c) 2015, Will Squire. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

// TODO import StreamTrigger as streamTrigger, so this will be streamTrigger.TriggerHandle

library inject.base;

import 'dart:io';
import 'dart:async';
import 'dart:convert';
import 'dart:mirrors'; // TODO This need replacing with a client side friendly reflection package
import 'package:stream_trigger/stream_trigger.dart';

class InjectModule
{
  //-------------------------------------------------------------------------------------------
  // Variables - Public
  //-------------------------------------------------------------------------------------------

  Map<String,Object> get dataMap => _dataMap;
  File get file => _file;
  String get prefix => _prefix;
  String get suffix => _suffix;

  //-------------------------------------------------------------------------------------------
  // Variables - Private
  //-------------------------------------------------------------------------------------------

  Map<String,Object> _dataMap;
  Map<List<int>,Object> _utf8DataMap;
  File _file;
  StreamTrigger _streamTrigger;
  String _prefix;
  String _suffix;
  List<int> _utfPrefix;
  List<int> _utfSuffix;
  StreamController _controller;
  StreamSubscription _subscription;
  List<Stream<List<int>>> _streams; // Streams queue
  bool _sync = true;
  bool _cancelOnError = true;
  //List<String> ignoredCharacters = [' ']; // Ignore spaces by default

  //-------------------------------------------------------------------------------------------
  // Functions - Constructor
  //-------------------------------------------------------------------------------------------

  /**
   * Maps values from Map keys to data in a module
   *
   * URI is the location of the resource module to be injected into and DataMap
   * is the map object of which the key names will be searched in
   * the target URI for the value to be 'injected' at these points
   */
  InjectModule(File this._file, Map<String,Object> this._dataMap, [String this._prefix, String this._suffix])
  {
    _initialiseTriggers();
    _initialiseStream();
  }

  //-------------------------------------------------------------------------------------------

  /**
   * Maps values from data model variables to data in a module
   *
   * URI is the location of the resource module to be injected into and the data
   * model is the object of which the property names will be searched in the
   * target URI for the value to be 'injected' at these points
   */
  InjectModule.fromObject(File this._file, Object dataModel, [String this._prefix, String this._suffix])
  {
    _dataMap = _objectToMap(dataModel);
    _initialiseTriggers();
    _initialiseStream();
  }

  //-------------------------------------------------------------------------------------------
  // Functions - Public
  //-------------------------------------------------------------------------------------------

  /**
   * Process
   *
   * Returns the the stream this class outputs.
   */
  Stream<List<int>> process()
  {
    return _controller.stream;
  }

  //-------------------------------------------------------------------------------------------
  // Functions - Private
  //-------------------------------------------------------------------------------------------

  /**
   * Initialise data map
   *
   * Creates new key/value pair map of pattern/data to map to occurring
   * pattern.
   */
  void _initialiseTriggers()
  {
    _utf8DataMap = _stringKeysToUTF8(_dataMap);

    if (_prefix != null)
    {
      _utfPrefix = UTF8.encode(_prefix);

      if (_suffix != null)
        _utfSuffix = UTF8.encode(_suffix);
    }

    _streamTrigger = new StreamTrigger(_resetTriggerHandle());
  }

  //-------------------------------------------------------------------------------------------

  /**
   * Set reset trigger handles
   *
   * Wrapper object essentially, adds the reset trigger handle
   * to StreamTrigger and always returns null as not to inject
   * into stream.
   */
  Null _setResetTriggerHandles()
  {
    _streamTrigger.setTriggers(_resetTriggerHandle());

    return null;
  }

  //-------------------------------------------------------------------------------------------

  /**
   * Reset handler
   *
   * Returns the handler for resetting trigger handles back
   * to the default state.
   */
  Map<List<int>,TriggerHandle> _resetTriggerHandle()
  {
    Map<List<int>,TriggerHandle> triggerHandles = new Map<List<int>,TriggerHandle>();

    if (_prefix != null)
    {
      // Add prefix trigger
      triggerHandles.addAll(_prefixTriggerHandle());
    }
    else
    {
      // Add command triggers
      triggerHandles.addAll(_injectHandles());
    }

    return triggerHandles;
  }

  //-------------------------------------------------------------------------------------------

  /**
   * Prefix trigger
   *
   * Returns a Map with the prefix trigger and handler.
   */
  Map<List,TriggerHandle> _prefixTriggerHandle()
  {
    return {
        _utfPrefix : ()
        {
          Map<List<int>,TriggerHandle> triggerHandles = new Map<List<int>,TriggerHandle>();

          // Adds itself in case one trigger disables this, then the prefix trigger occurs during contested
          triggerHandles.addAll(_prefixTriggerHandle());

          if (_suffix != null)
          {
            // Add command and suffix triggers (allow spaces)
            triggerHandles.addAll(_injectHandles());
            triggerHandles.addAll(_suffixTriggerHandle());
          }
          else
          {
            // Add command and space triggers (don't allow spaces and revert when command complete)
            triggerHandles.addAll(_injectTriggerHandlesWithReset());
            triggerHandles.addAll(_spaceResetTrigger());
          }

          _streamTrigger.setTriggers(triggerHandles);

          return null;
        }
    };
  }

  //-------------------------------------------------------------------------------------------

  /**
   * Suffix trigger
   *
   * Returns a Map with the suffix trigger and handler.
   */
  Map<List,TriggerHandle> _suffixTriggerHandle()
  {
    return {
        _utfSuffix : () => _setResetTriggerHandles()
    };
  }

  //-------------------------------------------------------------------------------------------

  /**
   * Initialise Stream
   *
   * Creates a stream that outputs the results of all different
   * files and compiles in a sequential fashion.
   */
  void _initialiseStream()
  {
    _streams = new List<Stream<List<int>>>();

    _addStream(
        _file.openRead()
        .transform(_streamTrigger)
    );

    _controller = new StreamController<List<int>>
    (
        onListen: _listen,
        onCancel: _onCancel,
        onPause: () { _subscription.pause(); },
        onResume: () { _subscription.resume(); },
        sync: _sync
    );
  }

  //-------------------------------------------------------------------------------------------

  /**
   * Trigger handles
   *
   * Returns triggers and corresponding handles from the utf data map.
   */
  Map<List<int>,TriggerHandle> _injectHandles()
  {
    Map<List<int>,TriggerHandle> triggerHandles = new Map<List<int>,TriggerHandle>();

    _utf8DataMap.forEach((List<int> k, Object v)
    {
      /*
      switch (v.runtimeType)
      {
        case InjectModule:
          triggerHandles[k] = () => v.process();
          break;
        case Stream:
          triggerHandles[k] = () => v;
          break;
        case String:
          triggerHandles[k] = () => UTF8.encode(v);
          break;
        default:
          triggerHandles[k] = () => UTF8.encode(v.toString());
      }
      */

      if (v is InjectModule)
        triggerHandles[k] = () => v.process();
      else if (v is String)
        triggerHandles[k] = () => UTF8.encode(v);
      else if (v is Stream<List<int>>)
        triggerHandles[k] = () => v;
      else
        triggerHandles[k] = () => UTF8.encode(v.toString());

    });

    return triggerHandles;
  }

  //-------------------------------------------------------------------------------------------

  /**
   * Inject trigger handles with reset
   *
   * Returns triggers and corresponding handles from the utf data map,
   * the 'trigger' is what pattern/sequence of element/s are to be
   * detected on the stream and the handler is callback function is
   * called once this occurs. This also resets StreamTrigger's trigger
   * handles after triggering.
   */
  Map<List<int>,TriggerHandle> _injectTriggerHandlesWithReset()
  {
    Map<List<int>,TriggerHandle> triggerHandles = new Map<List<int>,TriggerHandle>();

    _utf8DataMap.forEach((List<int> k, Object v)
    {
      /*
      switch (v.runtimeType)
      {
        case InjectModule:
          triggerHandles[k] = ()
          {
            _setResetTriggerHandles();
            return v.process();
          };
          break;
        case Stream:
          triggerHandles[k] = ()
          {
            _setResetTriggerHandles();
            return v;
          };
          break;
        case String:
          triggerHandles[k] = ()
          {
            _setResetTriggerHandles();
            return UTF8.encode(v);
          };
          break;
        default:
          triggerHandles[k] = ()
          {
            _setResetTriggerHandles();
            return UTF8.encode(v.toString());
          };
      }
      */
      if (v is InjectModule)
        triggerHandles[k] = ()
        {
          _setResetTriggerHandles();

          return v.process();
        };
      else if (v is String)
        triggerHandles[k] = ()
        {
          _setResetTriggerHandles();

          return UTF8.encode(v);
        };
      else if (v is Stream<List<int>>)
        triggerHandles[k] = ()
        {
          _setResetTriggerHandles();

          return v;
        };
      else
        triggerHandles[k] = ()
        {
          _setResetTriggerHandles();

          return UTF8.encode(v.toString());
        };
    });

    return triggerHandles;
  }

  //-------------------------------------------------------------------------------------------

  /**
   * Space reset trigger
   *
   * Returns trigger and corresponding handler for
   * resetting to initial state upon a space.
   */
  Map<List<int>,TriggerHandle> _spaceResetTrigger()
  {
    return {
        UTF8.encode(" ") : () => _setResetTriggerHandles()
    };
  }

  //-------------------------------------------------------------------------------------------

  /**
   * Add new stream
   *
   * Add a new stream reading from the selected file (streams can
   * be paused and others can be fed in instead).
   */
  void _addStream(Stream<List<int>> s)
  {
    _streams.add(s);
  }

  //-------------------------------------------------------------------------------------------

  /**
   * Listen to next stream
   *
   * Try to listen to the next stream if present. If no other streams
   * are present, close this outputting stream.
   */
  void _listen()
  {
    _subscription = _streams.first.listen
    (
        _controller.add,
        onError: _controller.addError,
        onDone: _done,
        cancelOnError: _cancelOnError
    );
  }

  //-------------------------------------------------------------------------------------------

  /**
   * Input stream is done
   *
   * Check for other input streams to listen to, else if non exist close
   * the outputted stream also.
   */
  void _done()
  {
    _streams.removeAt(0);

    if (_streams.length == 0)
    {
      //_subscription = null;
      _controller.close();
    }
    else
    {
      _listen();
    }
  }

  //-------------------------------------------------------------------------------------------

  /**
   * On cancelling listening to this stream
   *
   * Cancel and remove all subscriptions that only this stream
   * depends on.
   */
  void _onCancel()
  {
    _subscription.cancel();
    _subscription = null;
    _streams.clear();
  }

  //-------------------------------------------------------------------------------------------

  /**
   * Creates a Map from an Object
   *
   * Uses refection (mirrors) to produce a Map (array) from an object's
   * variables. Making the variable name the key, and it's value the
   * value.
   */
  Map<String,Object> _objectToMap(Object object)
  {
    // Mirror the particular instance (rather than the class itself)
    InstanceMirror instanceMirror = reflect(object);
    Map<String,String> dataMapped = new Map<String,String>();

    // Mirror the instance's class (type) to get the declarations
    for (var declaration in instanceMirror.type.declarations.values)
    {
      // If declaration is a type of variable, map variable name and value
      if (declaration is VariableMirror)
      {
        String variableName = MirrorSystem.getName(declaration.simpleName);
        Object variableValue = instanceMirror.getField(declaration.simpleName).reflectee;

        dataMapped[variableName] = variableValue;
      }
    }

    return dataMapped;
  }

  //-------------------------------------------------------------------------------------------

  /**
   * Encodes Map keys in UTF-8 encoding
   *
   * Converts a Map using Strings as keys into a new Map that
   * encodes the keys into UTF-8 Int types. Useful for streams
   * as both the data on the stream and the Map lookup will be in
   * the same format. Note, if Map data is not of the String type,
   * super class's toString() is called to convert to string.
   */
  Map<List<int>,Object> _stringKeysToUTF8(Map<String,Object> stringMap)
  {
    Map<List<int>,Object> utfMap = new Map<List<int>,Object>();

    stringMap.forEach((String key, Object value)
    {
      utfMap[UTF8.encode(key)] = value;
    });

    return utfMap;
  }
}

example/inject.dart
// Copyright (c) 2015, Will Squire. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

library inject.example;

import 'package:inject_module/inject_module.dart';
import 'dart:convert';
import 'dart:io';

main() {
  exampleOne();
  exampleTwo();
  exampleThree();
}

class User {
  // Example class. Passing this class as an inject data map
  // will mean 'name' and 'age' are the triggers, and their
  // values are what will be injected upon triggering.
  String name = 'Will Squire';
  int age = 25;
}

exampleOne() {
  // Specify the location of the file Inject will analyse
  // and process.
  File file = new File('../LICENSE');

  // Define either an Object or a Map to use as Inject's data
  // map. Inject uses the data maps' keys/variable names as
  // the triggers to replace/inject data. Once a trigger is
  // found in the source file's data, its replaced with the
  // data maps' corresponding value for the trigger.
  User user = new User();

  // Inject requires at least two arguments, one is the source
  // file and the other is the data map. Inject can also take
  // another two optional parameters, a prefix and then a suffix.
  // Using a prefix or both a prefix and suffix (circumfix)
  // means inject triggers need these additional significations
  // before triggers can be detected in the data. If using an
  // object as the data map, call fromObject constructor.
  InjectModule injectModule = new InjectModule.fromObject(file, user);

  // Inject can now analyse and transform the given file
  // data, returning the output as a stream.
  injectModule.process()
  .transform(UTF8.decoder)
  .listen((String char) {
    // Do something with the compiled result.
    print(char);
  });
}

exampleTwo() {
  // Defining the dataMap as a map type. In this data map the
  // trigger 'body' in will inject the content of a file as a
  // stream. Thus, injecting one stream into another.
  Map<String,Object> dataMap = {
    'body' : new File(new Uri.file('../README.md').toFilePath()).openRead()
  };

  // A third parameters has been passed here, which means
  // triggers will need the prefix of '$' to be recognised, and
  // have the corresponding value 'injected' in its place.
  // (Note, because Dart uses '$' for string interpolation,
  // it has been escaped with a '\').
  InjectModule injectModule = new InjectModule(new File('../LICENSE'), dataMap, '\$');

  // Process
  injectModule.process()
  .transform(UTF8.decoder)
  .listen((String char) {
    // Result.
    print(char);
  });
}

exampleThree() {
  // The third and forth parameters have been passed for each
  // Inject module here, which means triggers will need the
  // prefix of '<!--' and suffix of '-->' to be recognised, to
  // have the corresponding value 'injected' in its place.
  // Notice that HTML comment syntax was used for this. This
  // is so the files can be used even without Inject processing
  // the trigger, as it will not show up on the client's
  // machine.
  String prefix = '<!--';
  String suffix = '-->';

  Map<String,Object> headerDataMap = {
    'title' : "Website title"
  };

  Map<String,Object> footerDataMap = {
    'contactNumber' : "0800-Etc"
  };

  // In this data map, the values given are other Inject modules.
  // This takes the output from those streams and injects them
  // at the given location when this stream is processed.
  Map<String,Object> dataMap = {
    'header' : new InjectModule(new File('../LICENSE'), headerDataMap, prefix, suffix),
    'footer' : new InjectModule(new File('../LICENSE'), footerDataMap, prefix, suffix)
  };

  InjectModule injectModule = new InjectModule(new File('../LICENSE'), dataMap, prefix, suffix);

  // Process
  injectModule.process()
  .transform(UTF8.decoder)
  .listen((String char) {
    // Result.
    print(char);
  });
}